Angular to react conversion Reconfigure form

akhilkr128 opened the pull request:
Before
----------
<img width="1560" alt="Screenshot 2023-03-13 at 2 49 03 PM" src="https://user-images.githubusercontent.com/16753936/225380474-e86a9053-f470-48b9-b75d-0f2f5bd897c2.png">


After
----------
<img width="1589" alt="Screenshot 2023-05-01 at 10 49 55 PM" src="https://user-images.githubusercontent.com/16753936/235495798-1909070e-fdd6-4010-907c-25cfe20105c8.png">


https://user-images.githubusercontent.com/16753936/235605331-c047ce10-9c96-4525-93c2-7355bf3b8821.mov



Links
----------------

Issue - https://github.com/ManageIQ/manageiq-ui-classic/issues/7603

Steps for Testing/QA
-------------------------------

Compute -> Infrastructure -> Virtual Machines -> Configuration -> Reconfigure VM


@miq-bot add_label enhancement

akhilkr128 commented:
All form snapshots were updated because of the <MiqFormRenderer /> component changes.

jeffibm commented:
few minor suggestions

1. could you add a `margin-bottom:10px` to both buttons - `Add Disk` and `Add Network`
2. could you remove extra spaces for switches inside the data tables? eg -` Delete Backing`.
3. could you add some additional style for the Edit and Delete buttons for the network adapters list to align right? The same applies for `Resize` and `Delete` for the disk list
4. should we remove the buttons in the list and use links instead? looks like there are too many buttons on the page..
5. would it be possible to put all of these sections in our `MiqStructuredList`?

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/87487049/237054277-42e371fc-022d-4913-b3a0-05e430c21e97.png">

jeffibm commented:
For the edit, resize and action headers styling

1. in app/stylesheet/miq-data-table.scss, put this under .miq-data-table {
```
.header-button {
      width: 100px;
    }
```
2. in app/javascript/components/miq-data-table/index.jsx, add className to `<TableHeader` in  `const renderHeaders`
`className={classNames('miq-data-table-header', (header.isButton ? 'header-button' : ''))}`

3. add `isButton:true` to headers with buttons for the 3 tables, like
`{ key: 'edit', header: __('Edit'), isButton: true }`

jeffibm commented on app/javascript/components/reconfigure-vm-form/index.jsx:
```diff
@@ -0,0 +1,270 @@
+      );
+    }
+  }
+  let canReset = true;
+  let hasCustomTemplate = true;
+  if (data.form.type !== 'reconfigure') {
+    canReset = false;
+    hasCustomTemplate = false;
+  }
```
can you move this to top section.

jeffibm commented on app/javascript/components/reconfigure-vm-form/index.jsx:
```diff
@@ -0,0 +1,270 @@
+    isLoading: true,
+    form: {
+      title: formData.reconfigure.title,
+      type: 'reconfigure',
```
could you use this like enums `TYPE.RECONFIGURE`
```
const TYPE = {
    RECONFIGURE: 'reconfigure',
  };
```

akhilkr128 replied:
done

jeffibm commented on app/javascript/components/reconfigure-vm-form/index.jsx:
```diff
@@ -0,0 +1,270 @@
+      clearOnUnmount={clearOnUnmount}
+      canSubmit={getFormSubmitStatus(data)}
+    >
+      {diskTable}
+      {networkTable}
+      {cdromTable}
```
can you try this with the `componentMapper` and include them in schema? (if possible)

jeffibm commented on app/javascript/components/reconfigure-vm-form/helpers/disk.js:
```diff
@@ -0,0 +1,351 @@
+export const setDiskData = (disks, roles, data, setData) => {
+  const diskData = [];
+  // eslint-disable-next-line no-plusplus
+  for (let i = 0; i < disks.length; i++) {
```
lets try to break the for loop and use forEach if possible

akhilkr128 replied:
done.

jeffibm commented on app/javascript/components/reconfigure-vm-form/network-table/index.jsx:
```diff
@@ -0,0 +1,36 @@
+}) => (
+  <div>
+    <hr />
+    <h3>{__('Network Adapters')}</h3>
```
I think its better to say `Add Network Adapters` or `Edit Network Adapters`

akhilkr128 replied:
Made the form title changes.

akhilkr128 commented on app/javascript/components/reconfigure-vm-form/disk-table/index.jsx:
```diff
@@ -0,0 +1,36 @@
+  <div>
+    <hr />
+    <h3>{__('Disks')}</h3>
+    <button className="bx--btn bx--btn--primary pull-right disk-add" id="addDisk" onClick={() => buttonClick('new', 'disk', 'add')} type="button">
```
done.

jeffibm commented on app/javascript/spec/reconfigure-vm-form/reconfigure-vm-form.spec.js:
```diff
@@ -0,0 +1,173 @@
+import { act } from 'react-dom/test-utils';
+import ReconfigureVmForm from '../../components/reconfigure-vm-form/index';
+import {
+  valueFromHelpers, valueFromHelpersTwo, responseDataOne, responseDataThree, responseDataTwo, responseDataFour,
```
seems like this variable is not being used.

akhilkr128 replied:
done.

jeffibm commented on app/javascript/components/reconfigure-vm-form/helpers/network.js:
```diff
@@ -0,0 +1,301 @@
+  if (data.submitParams.networkIds.add.includes(row.id)) {
+    // remove from the newly added networks
+    // eslint-disable-next-line no-alert
+    if (window.confirm(__('Are you sure you want to delete this entry?'))) {
```
maybe we can change the confirmation to carbon in a different PR

akhilkr128 replied:
Sure.

jeffibm commented on app/javascript/components/reconfigure-vm-form/reconfigure-form-fields.js:
```diff
@@ -0,0 +1,152 @@
+  label: __('Memory'),
+  onText: __('Yes'),
+  offText: __('No'),
+  autoFocus: true,
```
we can remove the `autofocus:true` since the setState is rendering the form and focus goes to the top of the page

jeffibm reviewed (approved):
Looks good to me. Great work @akhilkr128 .
However, would like to have an extra pair of eyes to have a quick test on this..
cc @DavidResende0

akhilkr128 commented:
@DavidResende0 

> The main reconfigure form is missing a reset button 

- added

> Both the memory size and type inputs can be moved to the same line as the toggle 

- I will address this layout change in a different PR.


> I think the buttons in the data tables could be smaller. This is more of a project wide enhancement that could be done in a separate pr.

 - Changed the button styles to link.

> The way the delete buttons in the table are hidden off screen is going to confuse people. I'm not 100% what the best way to address this is but something should be done before we merge.

 - Fixed the issue.

<img width="1099" alt="Screenshot 2023-06-28 at 5 39 27 PM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/16753936/6f2f6fcf-0715-4321-bb83-6c61b94a55e8">

<img width="1063" alt="Screenshot 2023-06-28 at 5 58 49 PM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/16753936/bb91e970-8851-446b-82a9-98a7c4a1d57f">


cc: @jeffibm

DavidResende0 reviewed (approved):
LGTM, only thing I'll mention is that right now in the disks table, the toggle is not lined up with the rest of the row. This is fine to address in the followup pr you mentioned though.
![image](https://github.com/ManageIQ/manageiq-ui-classic/assets/64800041/00c10181-a0d3-41b5-a47f-750ae4f76056)

DavidResende0 commented:
@Fryguy is the failing security check a concern?

jeffibm commented:
@akhilkr128 , could you rebase this once again.

Fryguy commented:
> is the failing security check a concern?

This PR doesn't seem to have introduced this change, so I wouldn't worry about it.  It would be different if this PR introduced the bad package.
